## Week 4, Lab 1: Predicting Left-Handedness from Psychological Factors
> Author: Matt Brems

We can sketch out the data science process as follows:
1. Define the problem.
2. Obtain the data.
3. Explore the data.
4. Model the data.
5. Evaluate the model.
6. Answer the problem.

We'll walk through a full data science problem in this lab. 

---
## Step 1: Define The Problem.

You're currently a data scientist working at a university. A professor of psychology is attempting to study the relationship between personalities and left-handedness. They have tasked you with gathering evidence so that they may publish.

Specifically, the professor says "I need to prove that left-handedness is caused by some personality trait. Go find that personality trait and the data to back it up."

As a data scientist, you know that any real data science problem must be **specific** and **conclusively answerable**. For example:
- Bad data science problem: "What is the link between obesity and blood pressure?"
    - This is vague and is not conclusively answerable. That is, two people might look at the conclusion and one may say "Sure, the problem has been answered!" and the other may say "The problem has not yet been answered."
- Good data science problem: "Does an association exist between obesity and blood pressure?"
    - This is more specific and is conclusively answerable. The problem specifically is asking for a "Yes" or "No" answer. Based on that, two independent people should both be able to say either "Yes, the problem has been answered" or "No, the problem has not yet been answered."
- Excellent data science problem: "As obesity increases, how does blood pressure change?"
    - This is very specific and is conclusively answerable. The problem specifically seeks to understand the effect of one variable on the other.

### 1. In the context of the left-handedness and personality example, what are three specific and conclusively answerable problems that you could answer using data science? 

> You might find it helpful to check out the codebook in the repo for some inspiration.

Answer:
1. As the liking for guns increases, how likely is a person left-handedness?
2. As preference for mathematics class to pottery increases, how likely is a person left-handedness?
3. As liking for dancing increases, how likely is a person left-handedness?

---
## Step 2: Obtain the data.

### 2. Read in the file titled "data.csv."
> Hint: Despite being saved as a .csv file, you won't be able to simply `pd.read_csv()` this data!

In [1]:
#import libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.linear_model import Lasso, LassoCV, Ridge
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import Pipeline

from imblearn.over_sampling import SMOTE
from imblearn.pipeline import Pipeline
import warnings
warnings.filterwarnings('ignore')

In [2]:
#read file
df = pd.read_csv('./data.csv', sep='\t')
df.head()

Q1  Q2  Q3  Q4  Q5  Q6  Q7  Q8  Q9  Q10  ...  country  fromgoogle  engnat  \
0   4   1   5   1   5   1   5   1   4    1  ...       US           2       1   
1   1   5   1   4   2   5   5   4   1    5  ...       CA           2       1   
2   1   2   1   1   5   4   3   2   1    4  ...       NL           2       2   
3   1   4   1   5   1   4   5   4   3    5  ...       US           2       1   
4   5   1   5   1   5   1   5   1   3    1  ...       US           2       1   

   age  education  gender  orientation  race  religion  hand  
0   22          3       1            1     3         2     3  
1   14          1       2            2     6         1     1  
2   30          4       1            1     1         1     2  
3   18          2       2            5     3         2     2  
4   22          3       1            1     3         2     3  

[5 rows x 56 columns]

### 3. Suppose that, instead of us giving you this data in a file, you were actually conducting a survey to gather this data yourself. From an ethics/privacy point of view, what are three things you might consider when attempting to gather this data?
> When working with sensitive data like sexual orientation or gender identity, we need to consider how this data could be used if it fell into the wrong hands!

Answer:
1. Do not include personalized data such as actual name, phone number and addresses

---
## Step 3: Explore the data.

### 4. Conduct exploratory data analysis on this dataset.
> If you haven't already, be sure to check out the codebook in the repo, as that will help in your EDA process.

In [3]:
# Check for nulls
df.isna().sum()

Q1             0
Q2             0
Q3             0
Q4             0
Q5             0
Q6             0
Q7             0
Q8             0
Q9             0
Q10            0
Q11            0
Q12            0
Q13            0
Q14            0
Q15            0
Q16            0
Q17            0
Q18            0
Q19            0
Q20            0
Q21            0
Q22            0
Q23            0
Q24            0
Q25            0
Q26            0
Q27            0
Q28            0
Q29            0
Q30            0
Q31            0
Q32            0
Q33            0
Q34            0
Q35            0
Q36            0
Q37            0
Q38            0
Q39            0
Q40            0
Q41            0
Q42            0
Q43            0
Q44            0
introelapse    0
testelapse     0
country        0
fromgoogle     0
engnat         0
age            0
education      0
gender         0
orientation    0
race           0
religion       0
hand           0
dtype: int64

In [4]:
# Check column data types
df.dtypes

Q1              int64
Q2              int64
Q3              int64
Q4              int64
Q5              int64
Q6              int64
Q7              int64
Q8              int64
Q9              int64
Q10             int64
Q11             int64
Q12             int64
Q13             int64
Q14             int64
Q15             int64
Q16             int64
Q17             int64
Q18             int64
Q19             int64
Q20             int64
Q21             int64
Q22             int64
Q23             int64
Q24             int64
Q25             int64
Q26             int64
Q27             int64
Q28             int64
Q29             int64
Q30             int64
Q31             int64
Q32             int64
Q33             int64
Q34             int64
Q35             int64
Q36             int64
Q37             int64
Q38             int64
Q39             int64
Q40             int64
Q41             int64
Q42             int64
Q43             int64
Q44             int64
introelapse     int64
testelapse

In [5]:
df.shape

(4184, 56)

In [6]:
df.corr()['hand'].sort_values()

engnat        -0.051152
race          -0.046249
age           -0.037029
Q20           -0.031120
Q10           -0.030222
Q6            -0.026605
Q24           -0.024519
Q18           -0.022380
Q22           -0.021202
Q23           -0.019451
education     -0.018760
testelapse    -0.018266
Q14           -0.017737
Q8            -0.017541
introelapse   -0.017293
Q30           -0.016847
Q2            -0.014258
Q40           -0.013139
Q12           -0.008176
Q32           -0.005449
gender        -0.004415
Q21           -0.004328
Q36           -0.002710
Q41           -0.001979
Q16            0.000540
fromgoogle     0.005451
Q42            0.009638
religion       0.011911
Q34            0.012736
Q28            0.013155
Q19            0.013284
Q11            0.015954
Q15            0.016632
Q31            0.019128
Q13            0.021554
Q39            0.022652
Q37            0.025436
Q38            0.025876
Q44            0.027689
Q26            0.027984
Q9             0.028658
Q33            0

In [7]:
df['hand'].value_counts()

1    3542
2     452
3     179
0      11
Name: hand, dtype: int64

---
## Step 4: Model the data.

### 5. Suppose I wanted to use Q1 - Q44 to predict whether or not the person is left-handed. Would this be a classification or regression problem? Why?

Answer:

Classification because it is not trained with labeled data and Y is categorical

### 6. We want to use $k$-nearest neighbors to predict whether or not a person is left-handed based on their responses to Q1 - Q44. Before doing that, however, you remember that it is often a good idea to standardize your variables. In general, why would we standardize our variables? Give an example of when we would standardize our variables.

Answer: 
- Standardization is to make the scales of the variables even so that it is easier to interpret the data such as the coefficient
- Example: Height in meters versus weight in grams 

### 7. Give an example of when we might not standardize our variables.

Answer:

Height in meters versus weight in kilograms

### 8. Based on your answers to 6 and 7, do you think we should standardize our predictor variables in this case? Why or why not?

Answer: 
- There is no need for us to standardize our variables as they are all on the scale of one to five

### 9. We want to use $k$-nearest neighbors to predict whether or not a person is left-handed. What munging/cleaning do we need to do to our $y$ variable in order to explicitly answer this question? Do it.

Answer:

Filter out people who answered both-handed and 0

In [8]:
df2 = df[(df['hand']!=3)&(df['hand']!=0)]
df2.shape

(3994, 56)

### 10. The professor for whom you work suggests that you set $k = 4$. In this specific case, why might this be a bad idea?

Answer: 

There could be possibilities that there are the same number nearest points identified.

Example, two point A and two point B

However, this could be solved by setting the weight parameter in the kNN function

### 11. Let's *(finally)* use $k$-nearest neighbors to predict whether or not a person is left-handed!

> Be sure to create a train/test split with your data!

> Create four separate models, one with $k = 3$, one with $k = 5$, one with $k = 15$, and one with $k = 25$.

> Instantiate and fit your models.

In [9]:
# Set up X and y.
X = df2.drop(['hand', 'engnat', 'age', 'education', 'gender', 'fromgoogle', 'introelapse','testelapse','country'], axis = 'columns')
y = df2['hand']

In [10]:
# Split our data into training and testing sets.
X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    test_size = 0.2,
                                                    random_state = 42)

In [11]:
# Instantiate a StandardScaler + kNN pipeline.
pipe = Pipeline([
    ('ss', StandardScaler()),
    ('sampling', SMOTE()), #using SMOTE because there is left-handed is a small number compared to right-handed
    ('knn', KNeighborsClassifier())])

In [12]:
#kf = KFold(n_splits=6, random_state=42, shuffle=True)
#        cv_results = cross_val_score(model, X_train_scaled, y_train, cv=kf)
#        results.append(cv_results)
        
#pipe = Pipeline([
#    ('ss', StandardScaler()),
#    ('knn', KNeighborsClassifier())
#])

In [13]:
# Define dictionary of hyperparameters.
pipe_params = {'knn__n_neighbors': [3, 5, 15,25]}

In [14]:
# Instantiate our GridSearchCV object.
cv = GridSearchCV(pipe, 
                  pipe_params,
                  scoring = 'recall_macro',
                  cv=5, 
                  verbose=1)

In [15]:
# Fit the GridSearchCV object to the data.
cv.fit(X_train, y_train)

Fitting 5 folds for each of 4 candidates, totalling 20 fits


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('ss', StandardScaler()),
                                       ('sampling', SMOTE()),
                                       ('knn', KNeighborsClassifier())]),
             param_grid={'knn__n_neighbors': [3, 5, 15, 25]},
             scoring='recall_macro', verbose=1)

In [16]:
cv.cv_results_

{'mean_fit_time': array([0.10184698, 0.02895522, 0.03155904, 0.02978439]),
 'std_fit_time': array([0.13146848, 0.00558226, 0.0058343 , 0.00502494]),
 'mean_score_time': array([0.04128699, 0.05227952, 0.03849044, 0.04637146]),
 'std_score_time': array([0.00660289, 0.00384954, 0.00766087, 0.00352751]),
 'param_knn__n_neighbors': masked_array(data=[3, 5, 15, 25],
              mask=[False, False, False, False],
        fill_value='?',
             dtype=object),
 'params': [{'knn__n_neighbors': 3},
  {'knn__n_neighbors': 5},
  {'knn__n_neighbors': 15},
  {'knn__n_neighbors': 25}],
 'split0_test_score': array([0.51146384, 0.51267637, 0.49305556, 0.51344797]),
 'split1_test_score': array([0.53428131, 0.57782187, 0.54001323, 0.51521164]),
 'split2_test_score': array([0.52247205, 0.51541701, 0.5220727 , 0.51810349]),
 'split3_test_score': array([0.51097585, 0.54456895, 0.52604192, 0.52097149]),
 'split4_test_score': array([0.52002759, 0.51297255, 0.50969311, 0.52097149]),
 'mean_test_score': 

In [17]:
cv.best_score_

0.5326913506401462

In [18]:
cv.best_params_

{'knn__n_neighbors': 5}

In [19]:
# Evaluate the best model on the test data.
cv.score(X_test, y_test)

0.48983225193859786

Being good data scientists, we know that we might not run just one type of model. We might run many different models and see which is best.

### 12. We want to use logistic regression to predict whether or not a person is left-handed. Before we do that, let's check the [documentation for logistic regression in sklearn](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html). Is there default regularization? If so, what is it? If not, how do you know?

Answer: Yes

This class implements regularized logistic regression using the ‘liblinear’ library, ‘newton-cg’, ‘sag’, ‘saga’ and ‘lbfgs’ solvers. Note that regularization is applied by default. It can handle both dense and sparse input. Use C-ordered arrays or CSR matrices containing 64-bit floats for optimal performance; any other input format will be converted (and copied).

### 13. We want to use logistic regression to predict whether or not a person is left-handed. Before we do that, should we standardize our features?

Answer: split the test and train set first

### 14. Let's use logistic regression to predict whether or not the person is left-handed.


> Be sure to use the same train/test split with your data as with your $k$-NN model above!

> Create four separate models, one with LASSO and $\alpha = 1$, one with LASSO and $\alpha = 10$, one with Ridge and $\alpha = 1$, and one with Ridge and $\alpha = 10$. *(Hint: Be careful with how you specify $\alpha$ in your model!)*

> Instantiate and fit your models.

In [23]:
# Apply scaler
ss = StandardScaler()
ss.fit(X_train)
X_train_s = ss.transform(X_train)
X_test_s = ss.transform(X_test)

In [24]:
models = {"Lasso1": LogisticRegression(penalty='l1', C=1.0, solver='liblinear', class_weight='balanced'),
          "Lasso10": LogisticRegression(penalty='l1', C=10.0, solver='liblinear', class_weight='balanced'),
          "Ridge1": LogisticRegression(penalty='l2', C=1.0, class_weight='balanced'),
          "Ridge10": LogisticRegression(penalty='l2', C=10.0, class_weight='balanced')}
train_scores = []
test_scores = []
for model in models.values():
    model.fit(X_train_s, y_train)
    train_score = model.score(X_train_s, y_train)
    train_scores.append(train_score)
    test_score = model.score(X_test_s, y_test)
    test_scores.append(test_score)
print(f"train_scores = {train_scores}, test_score= {test_scores}")

train_scores = [0.5871674491392801, 0.5865414710485133, 0.5868544600938967, 0.5868544600938967], test_score= [0.5832290362953693, 0.5807259073842302, 0.5807259073842302, 0.5807259073842302]


---
## Step 5: Evaluate the model(s).

### 15. Before calculating any score on your data, take a step back. Think about your $X$ variable and your $Y$ variable. Do you think your $X$ variables will do a good job of predicting your $Y$ variable? Why or why not? What impact do you think this will have on your scores?

Answer:

No, because the X variables are opinionated and subjective.

### 16. Using accuracy as your metric, evaluate all eight of your models on both the training and testing sets. Put your scores below. (If you want to be fancy and generate a table in Markdown, there's a [Markdown table generator site linked here](https://www.tablesgenerator.com/markdown_tables#).)
- Note: Your answers here might look a little weird. You didn't do anything wrong; that's to be expected!

Answer:

|Model|KNN3|KNN5|KNN15|KNN25|Lasso1|Lasso10|Ridge1|Ridge10|
|---|---|---|---|---|---|---|---|---|
|Train Score|0.5198|0.5327|0.5182|0.5177|0.5872|0.5865|0.5869|0.5869|
|Test Score|0.4898|0.4898|0.4898|0.4898|0.5832|0.5832|0.5832|0.5832|

### 17. In which of your $k$-NN models is there evidence of overfitting? How do you know?

Answer: 

All the KNN models, because the training score is higher than the test score

### 18. Broadly speaking, how does the value of $k$ in $k$-NN affect the bias-variance tradeoff? (i.e. As $k$ increases, how are bias and variance affected?)

Answer: 

As k increases, bias decreases and variance increases

### 19. If you have a $k$-NN model that has evidence of overfitting, what are three things you might try to do to combat overfitting?

Answer:

1. Reduce the value of k
2. Use fewers x variables
3. Reduce the number of observations

### 20. In which of your logistic regression models is there evidence of overfitting? How do you know?

Answer:

All the logistic regression models, because the training score is higher than the test score

### 21. Broadly speaking, how does the value of $C$ in logistic regression affect the bias-variance tradeoff? (i.e. As $C$ increases, how are bias and variance affected?)

Answer:

As C increases, bias decreases and variance increases

### 22. For your logistic regression models, play around with the regularization hyperparameter, $C$. As you vary $C$, what happens to the fit and coefficients in the model? What do you think this means in the context of this specific problem?

In [22]:
models = {"Lasso0.1": LogisticRegression(penalty='l1', C=0.1, solver='liblinear', class_weight='balanced'),
          "Lasso0.5": LogisticRegression(penalty='l1', C=0.5, solver='liblinear', class_weight='balanced'),
          "Ridge0.1": LogisticRegression(penalty='l2', C=0.1, class_weight='balanced'),
          "Ridge0.5": LogisticRegression(penalty='l2', C=0.5, class_weight='balanced')}
train_scores = []
test_scores = []
for model in models.values():
    model.fit(X_train_s, y_train)
    train_score = model.score(X_train_s, y_train)
    train_scores.append(train_score)
    test_score = model.score(X_test_s, y_test)
    test_scores.append(test_score)
print(f"train_scores = {train_scores}, test_score= {test_scores}")

train_scores = [0.5812206572769953, 0.5849765258215962, 0.5884194053208138, 0.5871674491392801], test_score= [0.571964956195244, 0.5857321652065082, 0.5819774718397998, 0.5807259073842302]


Answer:

bias decreases and variance increases

### 23. If you have a logistic regression model that has evidence of overfitting, what are three things you might try to do to combat overfitting?

Answer:

1. Reduce C
2. Reduce number of observations
3. Reduce number of X variables

---
## Step 6: Answer the problem.

### 24. Suppose you want to understand which psychological features are most important in determining left-handedness. Would you rather use $k$-NN or logistic regression? Why?

Answer:

Logistic regression, because it is faster.

### 25. Select your logistic regression model that utilized LASSO regularization with $\alpha = 1$. Interpret the coefficient for `Q1`.

Answer:

In [26]:
Lasso1 = LogisticRegression(penalty='l1', C=1.0, solver='liblinear', class_weight='balanced')
Lasso1.fit(X_train_s, y_train)
Lasso1.score(X_train_s, y_train)
Lasso1.score(X_test_s, y_test)

print(Lasso1.coef_)

[[-0.00519215 -0.0660803  -0.03984091 -0.15129771  0.07415824  0.03119926
   0.05041006 -0.1726186  -0.11088864  0.11277128  0.05277872 -0.05839809
  -0.0519725   0.03525362 -0.0365563   0.05425127  0.06239997  0.0035045
  -0.07275001 -0.0588702  -0.06993021 -0.13759133 -0.08635385  0.01331784
   0.10666229  0.16280268  0.02023692  0.02603809  0.12908298  0.05825712
   0.047779   -0.09956104 -0.05706317 -0.03281603  0.05610226  0.00751105
  -0.04134506  0.12921188 -0.13309394 -0.09849929 -0.06273279 -0.07759045
  -0.16060327 -0.00135538  0.09655105 -0.01257824 -0.0264862 ]]


### 26. If you have to select one model overall to be your *best* model, which model would you select? Why?
- Usually in the "real world," you'll fit many types of models but ultimately need to pick only one! (For example, a client may not understand what it means to have multiple models, or if you're using an algorithm to make a decision, it's probably pretty challenging to use two or more algorithms simultaneously.) It's not always an easy choice, but you'll have to make it soon enough. Pick a model and defend why you picked this model!

Answer:

Lasso 10, because lowest biasness and overfitting

### 27. Circle back to the three specific and conclusively answerable questions you came up with in Q1. Answer one of these for the professor based on the model you selected!

Answer:

As a person agrees that he/she/they like danicng, it is 58.32% more likely that the person is left-handed

### BONUS:
Looking for more to do? Probably not - you're busy! But if you want to, consider exploring the following. (They could make for a blog post!)
- Create a visual plot comparing training and test metrics for various values of $k$ and various regularization schemes in logistic regression.
- Rather than just evaluating models based on accuracy, consider using sensitivity, specificity, etc.
- In the context of predicting left-handedness, why are unbalanced classes concerning? If you were to re-do this process given those concerns, what changes might you make?
- Fit and evaluate a generalized linear model other than logistic regression (e.g. Poisson regression).
- Suppose this data were in a `SQL` database named `data` and a table named `inventory`. What `SQL` query would return the count of people who were right-handed, left-handed, both, or missing with their class labels of 1, 2, 3, and 0, respectively? (You can assume you've already logged into the database.)